File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/DownloadJobItemStatus.java
package com.ustadmobile.lib.db.entities;

import java.util.LinkedList;
import java.util.List;

/**
 * This is used as a memory efficient summary of the status of a download
 *
 */
public class DownloadJobItemStatus {

    private int jobItemUid;

    private transient List<DownloadJobItemStatus> parents;

    private long contentEntryUid;

    private long bytesSoFar;

    private long totalBytes;

    private byte status;

    private transient List<DownloadJobItemStatus> children;

    public DownloadJobItemStatus() {

    }

    public DownloadJobItemStatus(DownloadJobItem item) {
        jobItemUid = (int)item.getDjiUid();
        contentEntryUid = item.getDjiContentEntryUid();
        bytesSoFar = item.getDownloadedSoFar();
        totalBytes = item.getDownloadLength();
    }

    public int getJobItemUid() {
        return jobItemUid;
    }

    public long getBytesSoFar() {
        return bytesSoFar;
    }

    public long getTotalBytes() {
        return totalBytes;
    }

    public void setJobItemUid(int jobItemUid) {
        this.jobItemUid = jobItemUid;
    }

    public void setContentEntryUid(long contentEntryUid) {
        this.contentEntryUid = contentEntryUid;
    }

    public void setBytesSoFar(long bytesSoFar) {
        this.bytesSoFar = bytesSoFar;
    }

    public void setTotalBytes(long totalBytes) {
        this.totalBytes = totalBytes;
    }

    public final void incrementTotalBytes(long increment) {
        totalBytes += increment;
    }

    public final void incrementBytesSoFar(long increment) {
        bytesSoFar += increment;
    }

    public byte getStatus() {
        return status;
    }

    public void setStatus(byte status) {
        this.status = status;
    }

    public List<DownloadJobItemStatus> getParents() {
        return parents;
    }

    public void addParent(DownloadJobItemStatus parent) {
        if(parents == null)
            parents = new LinkedList<>();

        parents.add(parent);
    }

    public void addChild(DownloadJobItemStatus child) {
        if(children == null)
            children = new LinkedList<>();

        children.add(child);
    }

    public List<DownloadJobItemStatus> getChildren() {
        return children;
    }

    public long getContentEntryUid() {
        return contentEntryUid;
    }

    @Override
    public int hashCode() {
        return jobItemUid;
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof DownloadJobItemStatus
                && ((DownloadJobItemStatus)o).jobItemUid == this.jobItemUid;
    }
}

File: lib-database-entities/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/DownloadJobItemStatus.kt
package com.ustadmobile.lib.db.entities

import kotlinx.serialization.Serializable
import kotlin.jvm.Transient

/**
 * This is used as a memory efficient summary of the status of a download
 *
 */
@Serializable
class DownloadJobItemStatus() {

    var jobItemUid: Int = 0

    @field:Transient
    var parents: MutableList<DownloadJobItemStatus>? = null
        private set

    var contentEntryUid: Long = 0

    var bytesSoFar: Long = 0

    var totalBytes: Long = 0

    var status: Int = 0

    @field:Transient
    var children: MutableList<DownloadJobItemStatus>? = null
        private set

    constructor(item: DownloadJobItem) : this() {
        jobItemUid = item.djiUid.toInt()
        contentEntryUid = item.djiContentEntryUid
        bytesSoFar = item.downloadedSoFar
        totalBytes = item.downloadLength
    }

    fun incrementTotalBytes(increment: Long) {
        totalBytes += increment
    }

    fun incrementBytesSoFar(increment: Long) {
        bytesSoFar += increment
    }

    fun addParent(parent: DownloadJobItemStatus) {
        if (parents == null)
            parents = mutableListOf()

        parents!!.add(parent)
    }

    fun addChild(child: DownloadJobItemStatus) {
        if (children == null)
            children = mutableListOf()

        children!!.add(child)
    }

    override fun hashCode(): Int {
        return jobItemUid
    }


    override fun equals(other: Any?): Boolean {
        return other != null && other is DownloadJobItemStatus && other.jobItemUid == this.jobItemUid
    }

    override fun toString() = "DownloadJobItemStatus DjiUid $jobItemUid ( $bytesSoFar / $totalBytes bytes) Status = $status"
}
